gh-badges/lib/badge-renderers.js

```diff
@@ -0,0 +1,596 @@
+        escapeXml(logo)
+      )}"/>`,
+    }
+  } else {
+    return {
+      hasLogo: false,
+      totalLogoWidth: 0,
+      renderedLogo: '',
+    }
+  }
+}
+
+function renderText({
+  leftMargin,
+  horizPadding = 0,
+  content,
+  verticalMargin = 0,
+  shadow = false,
+}) {
+  if (content.length) {
+    const textLength = preferredWidthOf(content)
+    const escapedContent = escapeXml(content)
+
+    const shadowMargin = 150 + verticalMargin
+    const textMargin = 140 + verticalMargin
+
+    const outTextLength = 10 * textLength
+    const x = 10 * (leftMargin + 0.5 * textLength + horizPadding)
+
+    let renderedText = ''
+    if (shadow) {
+      renderedText = `<text x="${x}" y="${shadowMargin}" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="${outTextLength}">${escapedContent}</text>`
+    }
+    renderedText += `<text x="${x}" y="${textMargin}" transform="scale(.1)" textLength="${outTextLength}">${escapedContent}</text>`
+
+    return {
+      renderedText,
+      width: textLength,
+    }
+  }
+  return { renderedText: '', width: 0 }
+}
+
+function renderLinks({
+  links: [leftLink, rightLink] = [],
+  labelWidth,
+  messageWidth,
+  height,
+}) {
+  leftLink = escapeXml(leftLink)
+  rightLink = escapeXml(rightLink)
+  const hasLeftLink = leftLink && leftLink.length
+  const hasRightLink = rightLink && rightLink.length
+  const leftLinkWidth = hasRightLink ? labelWidth : labelWidth + messageWidth
+
+  function render({ link, width }) {
+    return `<a target="_blank" xlink:href="${link}"><rect width="${width}" height="${height}" fill="rgba(0,0,0,0)"/></a>`
+  }
+
+  return (
+    (hasRightLink
+      ? render({ link: rightLink, width: labelWidth + messageWidth })
+      : '') +
+    (hasLeftLink ? render({ link: leftLink, width: leftLinkWidth }) : '')
+  )
+}
+
+function renderBadge({ links, leftWidth, rightWidth, height }, main) {
+  const width = leftWidth + rightWidth
+  return `
+    <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${width}" height="${height}">
+    ${main}
+    ${renderLinks({ links, leftWidth, rightWidth, height })}
+    </svg>`
+}
+
+function stripXmlWhitespace(xml) {
+  return xml
+    .replace(/>\s+/g, '>')
+    .replace(/<\s+/g, '<')
+    .trim()
+}
+
+class Badge {
+  static get fontFamily() {
+    throw new Error('Not implemented')
+  }
+
+  static get height() {
+    throw new Error('Not implemented')
+  }
+
+  static get verticalMargin() {
+    throw new Error('Not implemented')
+  }
+
+  static get shadow() {
+    throw new Error('Not implemented')
+  }
+
+  constructor({
+    label,
+    message,
+    links,
+    logo,
+    logoWidth,
+    logoPadding,
+    color = '#4c1',
+    labelColor = '#555',
+  }) {
+    const { hasLogo, totalLogoWidth, renderedLogo } = renderLogo({
+      logo,
+      logoWidth,
+      logoPadding,
+    })
+    const hasLabel = label.length
+
+    labelColor = hasLabel || (hasLogo && labelColor) ? labelColor : color
```

Comment:
This is fairly puzzling, particularly the `(hasLogo && labelColor)` bit.

Reply:
Yeah its pretty weird isn't it. Again I can't link to a line here because GitHub won't render the files, but if you look at line 12 of https://github.com/badges/shields/blob/v3-dev/gh-badges/templates/flat-template.svg that's where it originally came from. Any idea what the original intent of that is?

Reply:
Maybe it was meant to be `(hasLabel || hasLogo) && labelColor`, which would make more sense.

Reply:
ccdc468 - on close inspection, checking `labelColor` is redundant here anyway